### ISSUE TITLE
How to document C extensions?

### ISSUE THREAD

**Issue by p-mongo** [lsegal/yard#1281]
Yard appears to have support for documenting C extension functions (e.g. https://github.com/lsegal/yard/blob/master/lib/yard/parser/c/c_parser.rb) but I am not able to find any documentation regarding how this works.  I looked through https://rubydoc.info/gems/yard/file/README.md and other pages there and all examples I see are given in Ruby.

https://yardoc.org/features.html also does not mention ability to document C extensions, which I would think is an important feature to include in the feature list.

How does one document C extensions with Yard?

**Comment by lsegal** [comment]
YARD supports the same syntax as RDoc for .c files. You can look at examples in Ruby source but it doesn't seem like there is any coherent documentation on this format.

Example:  https://github.com/ruby/ruby/blob/trunk/array.c#L945

Note that you can use the YARD formatting for the Docstring itself rather than the call-seq style directives, but comments need to be placed as per the example.

**Comment by p-mongo** [comment]
OK, I am making slow & painful progress on figuring this out by scavenging around the Ruby source and trying what I am seeing. I haven't been successful with the example you referenced for instance but I got some docs rendered by putting comments into function bodies in init.

Is there any interest in having user documentation for how to use yard? Does yard employ parsers maintained by Ruby core or does it reimplement parsers that Ruby also has? What is the proper place for the documentation on how to write the documentation that yard will accept?

**Comment by lsegal** [comment]
> Is there any interest in having user documentation for how to use yard?

YARD has a [Getting Started](https://rubydoc.info/gems/yard/file/docs/GettingStarted.md) guide that covers the basics of using YARD syntax to document Ruby source. More documentation is always welcome if you want to describe the abstract documentation concepts in Ruby itself.

> Does yard employ parsers maintained by Ruby core or does it reimplement parsers that Ruby also has?

YARD uses Ripper, which Ruby integrates with for getting at the AST. The rest of it is all custom. RDoc implements its own layer on top of Ripper (AFAIK) if you want to look at another implementation.

> What is the proper place for the documentation on how to write the documentation that yard will accept?

The Getting Started guide would be a good place but if you want to create a separate even more basic "Writing documentation in Ruby" style doc that could be a way to do it too!

**Comment by lsegal** [comment]
I should add here though:

Documenting CRuby support isn't fully supported; specifically it works, but it's not prioritized. The support is primarily meant for Ruby core documentation and legacy C libraries, so the goal is only parity for Ruby core. In other words, if Ruby core does it, we will probably support it, otherwise we probably won't. 

Realistically, the cleaner approach to building well documented libraries with native code would be to wrap your native classes inside of .rb files and document with standard Ruby style documentation.

I say this because a "Writing documentation in Ruby" guide would not be helpful if it spent considerable time on documenting .c files. An addendum could be useful, but more than that would be overkill given the amount of documentation being written for .c files.

**Comment by p-mongo** [comment]
> Realistically, the cleaner approach to building well documented libraries with native code would be to wrap your native classes inside of .rb files and document with standard Ruby style documentation.

Where can I read up on this?

**Comment by lsegal** [comment]
> Where can I read up on this?

You're operating in a bit of a dark space of the Ruby ecosystem by documenting CRuby code, so the best place to find information on this would actually be by example rather than explicit docs. As I mentioned before, there is actually very little codified information about how CRuby documentation should be built, conventions, etc, and there frankly is no specification on any of this.

Did a random search and found a reasonable example of this wrapping pattern via https://github.com/udzura/ruby-criu which exposes private methods that are exposed via public API in lib/* although there is no actual docs for this lib, but TLDR here is that you would build private generic functions that perform the low level system operations and expose the actual business logic in Ruby source. This has a side effect of keeping you from building business logic in C where perf isn't needed.

**Comment by p-mongo** [comment]
To clarify, I am looking to document C code. My interpretation of  https://github.com/udzura/ruby-criu/blob/master/ext/criu/criu.c is "the C code is trivial and does not need documentation", this does not apply to the code I am working with.

> the best place to find information on this would actually be by example rather than explicit docs

This has already occurred to me. I humbly submit that "best" in this sentence falls way short of "good enough".

**Comment by lsegal** [comment]
> I do not believe that "best" in this sentence reaches the standard of "good".

Sure, and I'm not disagreeing with this, but you're asking for what exists today, and that's what exists today. 

There is a great opportunity for someone (you?) to improve the state of CRuby documentation guides. If you're interested, that's a great problem to tackle, but note that that's a bit of a separable problem from YARD itself: YARD is just a tool to parse documentation from source using existing conventions for Ruby documentation-- documenting guides themselves are not necessarily tool specific (YARD, RDoc, etc) but rather convention specific, i.e., where should you put code comments, how should you organize supporting files?

That said,

> To clarify, I am looking to document C code.

I would argue that you should not be thinking about documenting "C code", but rather documenting the Ruby interface. If you're actually trying to document the underlying C code (structs, typedefs, etc), then YARD won't help you much. If you're trying to document the interface, there are many ways to tackle this, and the example above is what you should look at. Interpreting the example as "the C code is trivial and doesn't need docs" is a mistake, because the point isn't the lack of docs, but the structure used to expose the interface through Ruby code.

TLDR:

- If your code is pure C and you absolutely want to keep it this way, document C as per Ruby core standards-- every file in the Ruby repo is an example for you here. Look at array.c, or string.c, etc., there are plenty of documented classes modules and methods to get info from.
- If your code is wrapped by a Ruby interface (this is ideal), document the .rb files, not the internal C implementation. In this case your C code is effectively a "private implementation" and the user doesn't know that any of that source exists.

**Comment by p-mongo** [comment]
OK, I looked at this extension again:

- https://github.com/udzura/ruby-criu/blob/master/lib/criu.rb contains no code
- https://github.com/udzura/ruby-criu/blob/master/ext/criu/criu.c contains method definitions

My first thought is to put documentation next to method definitions, i.e. in the C file. I am not clear on what you are suggesting as the superior approach, could you elaborate?

**Comment by lsegal** [comment]
The intent of such a library is to document the Ruby API, i.e. https://github.com/udzura/ruby-criu/blob/14d7bc6fcf21c0acb0e7c2b32015d98f5db85154/lib/criu.rb#L9 is where you write docs. The fd operation is low level and private. Even though this is not correctly done (method isn't private, no actual docs), the intent here is to hide your impl behind Ruby code so that your API is documented as Ruby, not C.

**Comment by p-mongo** [comment]
We currently end up writing docstrings in the functions that sets up the extension init method: https://github.com/mongodb/bson-ruby/blob/master/ext/bson/init.c#L192

I would like the docstring to be attached to the method instead: https://github.com/mongodb/bson-ruby/blob/master/ext/bson/write.c#L367

There is no Ruby class that wraps the C functions in question. As I understood the suggestion is to put the docstring into some other file which is parsed by Ruby as Ruby code. This would move the documentation further away from the definition which I think is not helpful to have the documentation be *up to date*.

We do have a JRuby implementation of the referenced method [here](https://github.com/mongodb/bson-ruby/blob/master/src/main/org/bson/ByteBuf.java#L459) and we recognize that the source of truth for documentation of public API is the C extension.

**Comment by lsegal** [comment]
> There is no Ruby class that wraps the C functions in question. As I understood the suggestion is to put the docstring into some other file which is parsed by Ruby as Ruby code. 

The suggestion is to document Ruby code at the point of integration into the Ruby runtime. This way you can define contracts for your Ruby API are in Ruby code for Rubyists to read as Ruby, not C. You're deciding to do this at the C level, but as illustrated in the linked criu library, you don't need to expose your Ruby interface directly from C.

The suggestion gives you two options:

1. Do it like criu and separate your C interface from your Ruby interface. Note that in criu, the `CRIU.images_dir=` method is _not_ a method defined in C via `rb_define_method` even though it could be-- instead `images_dir=` is a "Pure Ruby" implementation that calls out to a _separate_ implementation written in C (`_set_images_dir_fd`). This is the [standard C++ "pimpl" pattern](https://www.geeksforgeeks.org/pimpl-idiom-in-c-with-examples/), probably better known as just interface hiding or 
2. If you really don't want to do this, yes, just make a separate .rb class that has no actual code in it, isn't loaded by any programs, but defines the interface. Yes, this will keep code separated from your impl, but I guess you brought up an interesting twist here:

> We do have a JRuby implementation of the referenced method here and we recognize that the source of truth for documentation of public API is the C extension.

So you kind of already have the problem of documentation far away from your contract, because you have multiple versions of this implementation all exposed in different ways to different environments, all doing the same thing with _copies of_ the same documentation. All of this reinforces the benefit of interface hiding, using private implementations for each platform, and documenting on the PUBLIC interface, i.e., the "Pure Ruby" methods that use these pimpls. Example:

```ruby
class BSON::ByteBuffer
  # Appends arguments to a buffer
  def push(*args)
    args.each do |arg|
      case arg
      when Integer
        BufferMethods.put_int32(self, arg)
      when String
        # write string as bytes
      end
    end
  end

  alias << push
end
```

And yes, I changed your interface significantly to show how interface hiding also gets you to much more idiomatic Ruby too (because `buf.put_int32(x); buf.put_int32(y)` isn't a useful Ruby interface IMO). You can of course just keep the plain old `put_int32` style methods and call out to your pimpl that way.

Doing this allows you to not have to write stringent public "Ruby" docs for your C code or your Java code. Those implementations can be documented for C or Java devs. Use javadoc or doxygen or whatever the right tool is for those devs, or simply just refer devs to the public contracts defined solely on the Ruby side.

Basically: YARD is not a tool primarily meant to document C or Java code, it's meant to document _Ruby interfaces_. I think the part where our lines are getting crossed is that you are treating your C code as a public Ruby interface, which Ruby can certainly allow you to do, but as mentioned, YARD does not fully support this from a documentation perspective because doing so would require us to spend more time parsing C than Ruby, and since that's not the goal of this project, such support will always lag behind actual Ruby specific features.

**Comment by p-mongo** [comment]
> I changed your interface significantly

I think this is the part that confuses me the most. In this issue I am asking, basically, how I am supposed to document my code using facilities that yard provides (however limited they may be). I am under the impression that _some_ facilities are provided that I can use, but I am having a hard time figuring out how to do so.

I can add a placeholder Ruby file and put the documentation there. The requirements for a solution like this are:

1. The documentation, when rendered, looks like other Ruby documentation, and integrates with other API documentation we have, some of which is covers Ruby code and is written in the "standard" way. So, for the method I referenced earlier, we need something that looks like https://api.mongodb.com/bson-ruby/current/BSON/ByteBuffer.html#put_int32-instance_method and is placed accordingly in the overall produced API documentation.
2. The file in which this documentation is placed does not conflict with the implementation. The implementation exists only as C or Java code.

I understand the idea behind implementing things in Ruby (Rubinius etc.) but this is not the codebase that we are working with. So merely saying "rewrite your library to have a  different architecture, then your documentation will work" doesn't seem helpful.

**Comment by lsegal** [comment]
> In this issue I am asking, basically, how I am supposed to document my code using facilities that yard provides (however limited they may be).

Okay now _I'm_ confused. If the question is "how do I do this?" then it looks to me like [you've already done it](https://api.mongodb.com/bson-ruby/current/BSON/ByteBuffer.html#put_int32-instance_method).

Looking at your code, I see blocks such as:

https://github.com/mongodb/bson-ruby/blob/master/ext/bson/init.c#L239-L247

Generating perfectly fine into docs such as:

![image](https://user-images.githubusercontent.com/686/80929722-1a417e80-8d63-11ea-952d-d82a9f1da7a5.png)

So I guess I'm not really sure what the question is? Looks like you already have it working in a way that YARD supports.

Is the question how to use YARD tags with C style comments?

**Comment by p-mongo** [comment]
I am able to generate documentation.

I would like to have the docstrings be in a different place: either next to the function definitions (https://github.com/mongodb/bson-ruby/blob/master/ext/bson/write.c#L367) or in a standalone file which only has the docstrings.

If I move the docblock you highlighted to write.c where the function is defined, it's not rendered in the documentation.

**Comment by lsegal** [comment]
> I would like to have the docstrings be in a different place: either next to the function definitions (https://github.com/mongodb/bson-ruby/blob/master/ext/bson/write.c#L367) or in a standalone file which only has the docstrings.

1. **Next to the functions**: Put docs [in the same .c file](https://github.com/ruby/ruby/blob/master/array.c#L3191-L3219) as the `rb_define_method` or if you use the [Document-method RDoc directive](https://github.com/ruby/ruby/blob/b99775b163ce44079c1f8727ce9b4ed8bb03489d/ext/openssl/ossl_bn.c#L366-L371). You can check that source file for plenty more RDoc syntaxes or see the [RDoc docs](https://docs.ruby-lang.org/en/2.1.0/RDoc/Parser/C.html) for how it expects this stuff to look. NB: YARD does not support all of these Document-* syntaxes.

2. **In a standalone file with just docstrings**: you can do this with `Document-method` syntax or just use Ruby files as the standalone files as suggested above:

```ruby
# separate_file.rb

class BSON::ByteBuffer
  # Documentation here
  def put_int64(value)
  end
end
```

**Comment by 0x1eef** [comment]
I stumbled across this issue, and what worked best for me - as of 2022, was to use the `@!method` directive. It was not practical for me to define the method in Ruby land, since that would conflict with the definition in C, so `@!method` was a good fit:

```ruby
# ffi.rb
module Lock::FFI
  ##
  # @!method lockf
  #   @param [Integer] fd
  #     The file descriptor.
  #   @param [Integer] cmd
  #     The cmd
  #   @param [Integer] len
  #     The number of bytes.
  #   @return [Integer]
  #     Returns 0 when successful.
  #   @see (https://man7.org/linux/man-pages/man3/lockf.3.html) lockf man page (Linux)
end
```
```c
// lockf.c
static VALUE
lockf_lock(VALUE self, VALUE fd, VALUE cmd, VALUE len) {
  ...
}
void Init_lockf(void) {
    rb_define_method(mFFI, "lockf", lockf_lock, 3);
}
```
